File: src/main/java/me/alegian/thavma/impl/client/gui/tooltip/AspectClientTooltipComponent.java
package me.alegian.thavma.impl.client.gui.tooltip;

import com.google.common.collect.ImmutableList;
import me.alegian.thavma.impl.client.T7GuiGraphics;
import me.alegian.thavma.impl.client.renderer.AspectRenderer;
import me.alegian.thavma.impl.common.aspect.AspectStack;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.screens.inventory.tooltip.ClientTooltipComponent;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.api.distmarker.OnlyIn;

@OnlyIn(Dist.CLIENT)
public class AspectClientTooltipComponent implements ClientTooltipComponent {
  private final ImmutableList<AspectStack> displayedAspects;
  private static final int WIDTH = AspectRenderer.PIXEL_RESOLUTION;
  private static final int PADDING = 3;

  public AspectClientTooltipComponent(AspectTooltipComponent tooltip) {
    displayedAspects = tooltip.getAspectMap().displayedAspects();
  }

  @Override
  public int getHeight() {
    if (isEmpty()) return 0;
    return WIDTH + PADDING * 2;
  }

  @Override
  public int getWidth(Font pFont) {
    if (isEmpty()) return 0;
    return (WIDTH + PADDING) * displayedAspects.size();
  }

  private boolean isEmpty() {
    return displayedAspects.isEmpty();
  }

  @Override
  public void renderImage(Font pFont, int pX, int pY, GuiGraphics pGuiGraphics) {
    if (isEmpty()) return;

    int i = 0;
    for (AspectStack aspectStack : displayedAspects) {
      AspectRenderer.INSTANCE.renderAspect(new T7GuiGraphics(pGuiGraphics), aspectStack, pX + i * (WIDTH + PADDING), pY + PADDING);
      i++;
    }
  }
}

File: src/main/java/me/alegian/thavma/impl/client/gui/tooltip/AspectClientTooltipComponent.kt
package me.alegian.thavma.impl.client.gui.tooltip

import me.alegian.thavma.impl.Thavma
import me.alegian.thavma.impl.client.renderer.AspectRenderer.renderAspect
import me.alegian.thavma.impl.client.util.translateXY
import me.alegian.thavma.impl.client.util.usePose
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.inventory.tooltip.ClientTooltipComponent

private const val WIDTH = 16
private const val PADDING = 3

class AspectClientTooltipComponent(tooltip: AspectTooltipComponent) : ClientTooltipComponent {
  private val aspects = tooltip.aspectMap

  override fun getHeight(): Int {
    if (aspects.size == 0) return 0
    return WIDTH + PADDING * 2
  }

  override fun getWidth(pFont: Font): Int {
    if (aspects.size == 0) return 0
    return (WIDTH + PADDING) * aspects.size
  }

  override fun renderImage(pFont: Font, pX: Int, pY: Int, guiGraphics: GuiGraphics) {
    for ((i, aspectStack) in aspects.withIndex()) {
      guiGraphics.usePose {
        translateXY(pX + i * (WIDTH + PADDING), pY + PADDING)
        renderAspect(guiGraphics, aspectStack)
      }
    }
  }

  object I18n {
    val NOT_SCANNED = "tooltip" + Thavma.MODID + ".not_scanned"
  }
}
